pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/SegmentSizeRule.java

```diff
@@ -0,0 +1,122 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.controller.recommender.rules.impl;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.io.File;
+import java.io.IOException;
+import org.apache.commons.io.FileUtils;
+import org.apache.pinot.controller.recommender.exceptions.InvalidInputException;
+import org.apache.pinot.controller.recommender.io.ConfigManager;
+import org.apache.pinot.controller.recommender.io.InputManager;
+import org.apache.pinot.controller.recommender.realtime.provisioning.MemoryEstimator;
+import org.apache.pinot.controller.recommender.rules.AbstractRule;
+import org.apache.pinot.controller.recommender.rules.io.configs.SegmentSizeRecommendations;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
+
+
+/**
+ * This rule generates a segment based on the provided schema characteristics and then recommends the followings
+ * using the size and number of records in the generated segments:
+ *   - number of segments
+ *   - number of records in each segment
+ *   - size of each segment
+ */
+public class SegmentSizeRule extends AbstractRule {
+
+  static final int MEGA_BYTE = 1024 * 1024;
+
+  public SegmentSizeRule(InputManager input, ConfigManager output) {
+    super(input, output);
+  }
+
+  @Override
+  public void run()
+      throws InvalidInputException {
+
+    if (_input.getTableType().equalsIgnoreCase("REALTIME")) {
+      // no need to estimate segment size & optimal number of segments for realtime only tables;
+      // RT Provisioning Rule will have a comprehensive analysis on that
+      return;
+    }
+
+    // generate a segment
+    TableConfig tableConfig = createTableConfig(_input.getSchema());
+    int numRowsInGeneratedSegment = _input.getSegmentSizeRuleParams().getNumRowsInGeneratedSegment();
+    File generatedSegmentDir =
+        new MemoryEstimator.SegmentGenerator(_input._schemaWithMetaData, _input._schema, tableConfig,
+            numRowsInGeneratedSegment, true).generate();
+
+    // estimate optimal segment count & size parameters
+    SegmentSizeRecommendations params = estimate(FileUtils.sizeOfDirectory(generatedSegmentDir),
+        _input.getSegmentSizeRuleParams().getDesiredSegmentSizeMb() * MEGA_BYTE, numRowsInGeneratedSegment,
+        _input.getNumRecordsPerPush());
+
+    // wire the recommendations
+    _output.setSegmentSizeRecommendations(params);
+    _input.capCardinalities((int) params.getNumRows());
+
+    // cleanup
+    try {
+      FileUtils.deleteDirectory(generatedSegmentDir);
+    } catch (IOException e) {
+      throw new RuntimeException("Cannot delete the generated segment directory", e);
+    }
+  }
+
+  /**
+   * Estimate optimal segment size parameters
```

Comment:
you may want to add a comment that this estimate is a rough one based on expectation of linear increase, which is typically not the case. Ideal way to do this is to iterate by generating new segments with expected number of rows, re-adjusting the ratio  until we find the segment size to be within tolerant limit (or a certain number of iterations are done)

Reply:
Good point. Added a comment.